graphql_compiler/schema_transformation/make_query_plan.py

```diff
@@ -0,0 +1,274 @@
+            )
+        ):
+            new_directives = copy(ast.directives)
+            new_directives.append(_get_in_collection_filter_directive(input_filter_name))
+            new_ast = copy(ast)
+            new_ast.directives = new_directives
+            return new_ast
+
+    if ast.selection_set is None:  # Nothing to recurse on
+        return ast
+
+    # Otherwise, recurse and look for field with desired out_name
+    made_changes = False
+    new_selections = []
+    for selection in ast.selection_set.selections:
+        new_selection = _add_filter_at_field_with_output(
+            selection, field_out_name, input_filter_name
+        )
+        if new_selection is not selection:  # Changes made somewhere down the line
+            if not made_changes:
+                made_changes = True
+            else:
+                # Change has already been made, but there is a new change. Implies that multiple
+                # fields have the @output directive with the desired name
+                raise GraphQLValidationError(
+                    u'There are multiple @output directives with the out_name "{}"'.format(
+                        field_out_name
+                    )
+                )
+        new_selections.append(new_selection)
+
+    if made_changes:
+        new_ast = copy(ast)
+        new_ast.selection_set = ast_types.SelectionSet(selections=new_selections)
+        return new_ast
+    else:
+        return ast
+
+
+def _is_output_directive_with_name(directive, out_name):
+    """Return whether or not the input is an @output directive with the desired out_name."""
+    if not isinstance(directive, ast_types.Directive):
+        raise AssertionError(u'Input "{}" is not a directive.'.format(directive))
+    return (
+        directive.name.value == OutputDirective.name and
+        directive.arguments[0].value.value == out_name
+    )
+
+
+def _get_in_collection_filter_directive(input_filter_name):
+    """Create a @filter directive with in_collecion operation and the desired variable name."""
+    return ast_types.Directive(
+        name=ast_types.Name(value=FilterDirective.name),
+        arguments=[
+            ast_types.Argument(
+                name=ast_types.Name(value='op_name'),
+                value=ast_types.StringValue(value='in_collection'),
+            ),
+            ast_types.Argument(
+                name=ast_types.Name(value='value'),
+                value=ast_types.ListValue(
+                    values=[
+                        ast_types.StringValue(value=u'$' + input_filter_name),
+                    ],
+                ),
+            ),
+        ],
+    )
+
+
+def print_query_plan(query_plan_descriptor):
+    """Return a string describing query plan."""
+    query_plan_str = u''
+    plan_and_depth = _get_plan_and_depth_in_dfs_order(query_plan_descriptor.root_sub_query_plan)
+
+    for query_plan, depth in plan_and_depth:
+        line_separation = u'\n' + u' ' * 8 * depth
```

Comment:
Let's use 4 spaces by default instead, and ideally make this configurable via a kwarg.

I'd make the function signature be `def print_query_plan(query_plan_descriptor, indentation_depth=4):`